contracts/airdrops.js

```diff
@@ -0,0 +1,244 @@
+
+  if (api.assert(isSignedWithActiveKey === true, 'you must use a custom_json signed with your active key')
+    && api.assert(symbol && typeof symbol === 'string'
+      && type && typeof type === 'string'
+      && list && typeof list === 'object' && list.length, 'invalid params')
+    && api.assert(type === 'transfer' || type === 'stake', 'invalid type')) {
+    const token = await api.db.findOneInTable('tokens', 'tokens', { symbol });
+
+    // get api.sender's utility and airdrop token balances
+    const { balance: utilityTokenBalance } = { ...await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: UTILITY_TOKEN_SYMBOL }) };
+    const { balance: nativeTokenBalance } = { ...await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol }) };
+
+    if (api.assert(token !== null, 'symbol does not exist')
+      && hasValidType(token, type)) {
+      const airdrop = await parseAirdrop(list, token.precision);
+
+      if (api.assert(airdrop.list.length > 0 && airdrop.isValid, 'invalid list')
+        && api.assert(utilityTokenBalance
+          && api.BigNumber(utilityTokenBalance).gte(airdrop.fee), 'you must have enough tokens to cover the airdrop fee')
+        && api.assert(nativeTokenBalance
+          && api.BigNumber(nativeTokenBalance).gte(airdrop.quantity), 'you must have enough tokens to do the airdrop')) {
+        // validations completed
+        // lock airdrop tokens by transfering them to contract
+        const tokenTransfer = await api.executeSmartContract('tokens', 'transferToContract', {
+          to: CONTRACT_NAME, symbol, quantity: airdrop.quantity,
+        });
+
+        if (transferIsSuccesfull(tokenTransfer, 'transferToContract', api.sender, CONTRACT_NAME, symbol, airdrop.quantity)) {
+          // deduct fee from sender's utility token balance
+          const feeTransfer = await api.executeSmartContract('tokens', 'transfer', {
+            to: 'null', symbol: UTILITY_TOKEN_SYMBOL, quantity: airdrop.fee, isSignedWithActiveKey,
+          });
+
+          if (transferIsSuccesfull(feeTransfer, 'transfer', api.sender, 'null', UTILITY_TOKEN_SYMBOL, airdrop.fee)) {
+            const res = await api.db.insert('pendingAirdrops', {
+              airdropId: api.transactionId,
+              symbol,
+              type,
+              list: airdrop.list,
+              blockNumber: api.blockNumber,
+            });
+
+            api.emit('newAirdrop', { airdropId: res.airdropId });
+          } else {
+            // if fee transfer was failed, return native balance to api.sender
+            await api.transferTokens(api.sender, symbol, airdrop.quantity, 'user');
+          }
+        }
+      }
+    }
+  }
+};
+
+const processAirdrop = async (airdrop, maxTransactionsPerBlock) => {
+  const {
+    airdropId,
+    list,
+    symbol,
+    type,
+  } = airdrop;
+
+  let airdropIsPending = true;
+  let count = 0;
+
+  const completedDrops = [];
+
+  while (airdropIsPending) {
```

Comment:
can shorten to  

while (count < maxTransactionsPerBlock && list[count] !== undefined) {
..
count += 1
}